### ISSUE TITLE
Displaying the form with the map

### ISSUE THREAD

**Issue by puziip** [ofrohn/d3-celestial#8]
Hi Olaf - great service firstly.

I am trying to add the form with the map (either form or location). I have tried to change these to true (rather than false) in the config variable I pass but I get an error. Am I missing something? Do I need to do something else?

Any help will be great. Primarily I just want to mainly use location like this demo of yours:

http://ofrohn.github.io/celestial-demo/location.html

Thanks
Paresh

**Comment by ofrohn** [comment]
What error message are you getting exactly? 
Also, what Browser are you using? Because recent versions of Firefox seem to have trouble with a relative path to the data files, like I do in the demo
```javascript
Celestial.display({
  location: true,
  projection: "airy",  
  datapath: "../data/",
  planets:  { show: true }
});
```
Alos set ``form: true`` if you want the full form. If you use Chrome or put the data in a subfolder ``datapath: 'data/'`` that should work.

**Comment by puziip** [comment]
Hi - thanks for getting back to me. This is the message:

celestial.js:2151 Uncaught TypeError: Cannot read property 'id' of null
    at enable (celestial.js:2151)
    at form (celestial.js:1943)
    at Object.Celestial.display (celestial.js:100)
    at d3.html:253

I feel I am missing something simple. This is the error when I try 'form: true'.
Equally this the error from setting location: true - 

celestial.js:3431 Uncaught TypeError: Cannot read property 'value' of null
    at daySel (celestial.js:3431)
    at new datetimepicker (celestial.js:3425)
    at geo (celestial.js:2315)
    at Object.Celestial.display (celestial.js:95)
    at d3.html:253

The browsers I am using is Chrome only - no firefox. My data path uses an Apache alias to get to the right directory on the web server before going into a sub directory with the data files. This works fine for the map without the form so I assume it is not the issue.

**Comment by ofrohn** [comment]
Ok, was just my first guess. 
I can't replicate the error. What may be the case is that the page with the map needs another tag like this:
``<div id="celestial-form"></div>``
to put the form & location inputs into. Do you have that?

**Comment by puziip** [comment]
Hi - yes I did. When I looked at your examples (and your celestial.css file) I realised I may need this tag. So it is there.

**Comment by ofrohn** [comment]
Well, damn, I can't replicate your error. What does the config look like that you're using?

**Comment by puziip** [comment]
Hmm. In that case here is pretty much all of the code - I have left form: true.

<!DOCTYPE html>
 
<html>
 
<head>
 
    <script type="text/javascript" src="/premap/jscss/lib/d3.min.js"></script>
    <script type="text/javascript" src="/premap/jscss/lib/d3.geo.projection.min.js"></script>
    <script type="text/javascript" src="/premap/jscss/celestial.js"></script>
    <script type="text/javascript" src="/premap/jscss/jquery-2.2.4.min.js"></script>
    <link rel="stylesheet" href="/premap/jscss/celestial.css">
 
</head>
 
<body>
 
<script type="text/javascript">
 
var config = {
  width: 1200,           // Default width, 0 = full parent element width; height is determined by projection
  projection: "lagrange",    // Map projection used: see below // Map projection used: airy, aitoff, armadillo, august, azimuthalEqualArea, azimuthalEquidistant, baker, berghaus, boggs, bonne, bromley, collignon, craig, craster, cylindricalEqualArea, cylindricalStereographic, eckert1, eckert2, eckert3, eckert4, eckert5, eckert6, eisenlohr, equirectangular, fahey, foucaut, ginzburg4, ginzburg5, ginzburg6, ginzburg8, ginzburg9, gringorten, hammer, hatano, healpix, hill, homolosine, kavrayskiy7 , lagrange, larrivee, laskowski, loximuthal, mercator, miller, mollweide, mtFlatPolarParabolic, mtFlatPolarQuartic, mtFlatPolarSinusoidal, naturalEarth, nellHammer, orthographic, patterson, polyconic, rectangularPolyconic, robinson, sinusoidal, stereographic, times, twoPointEquidistant, vanDerGrinten, vanDerGrinten2, vanDerGrinten3, vanDerGrinten4, wagner4, wagner6, wagner7, wiechel, winkel3
  transform: "equatorial", // Coordinate transformation: equatorial (default),
                           // ecliptic, galactic, supergalactic
  center: null,       // Initial center coordinates in set transform
                      // [longitude, latitude, orientation] all in degrees
                      // null = default center [0,0,0]
  orientationfixed: true,  // Keep orientation angle the same as center[2]
  geopos: null,       // optional initial geographic position [lat,lon] in degrees, overrides center
  adaptable: true,    // Sizes are increased with higher zoom-levels
  interactive: true,  // Enable zooming and rotation with mousewheel and dragging
  form: true,         // Display form for interactive settings
  location: false,    // Display location settings (no center setting on form)
  controls: false,     // Display zoom controls
  lang: "",           // Language for names, so far only for constellations:
                      // de: german, es: spanish. Default:en or empty string for english
  container: "#celestial-map",   // ID of parent element, e.g. div, null = html-body
  datapath: "/premap/jscss/data/",  // Path/URL to data files, empty = subfolder 'data'
  stars: {
    show: false,    // Show stars
    limit: 6,      // Show only stars brighter than limit magnitude
    colors: true,  // Show stars in spectral colors, if not use default color
    style: { fill: "#ffffff", opacity: 1 }, // Style for stars
    names: true,   // Show star designation (Bayer, Flamsteed, Variable star, Gliese, whichever applies first in that order)
    proper: false, // Show proper name (if one exists)
    desig: false,  // Show all designations, including Draper and Hipparcos
    namelimit: 2.5,  // Show only names/designations for stars brighter than namelimit
    namestyle: { fill: "#ddddbb", font: "11px Georgia, Times, 'Times Roman', serif", align: "left", baseline: "top" },  // Style for star designations
    propernamestyle: { fill: "#ddddbb", font: "11px Georgia, Times, 'Times Roman', serif", align: "right", baseline: "bottom" }, // Styles for star names
    propernamelimit: 1.5,  // Show proper names for stars brighter than propernamelimit
    size: 7,       // Maximum size (radius) of star circle in pixels
    exponent: -0.28, // Scale exponent for star size, larger = more linear
    data: 'stars.6.json' // Data source for stellar data,
                         // number indicates limit magnitude
  },
  dsos: {
    show: false,    // Show Deep Space Objects
    limit: 10,      // Show only DSOs brighter than limit magnitude
    names: true,   // Show DSO names
    desig: true,   // Show short DSO names
    namestyle: { fill: "#cccccc", font: "11px Helvetica, Arial, serif", align: "left", baseline: "top" }, // Style for DSO names
    namelimit: 6,  // Show only names for DSOs brighter than namelimit
    size: null,    // Optional seperate scale size for DSOs, null = stars.size
    exponent: 1.4, // Scale exponent for DSO size, larger = more non-linear
    data: 'dsos.bright.json', // Data source for DSOs, opt. number indicates limit magnitude
    symbols: {  //DSO symbol styles, 'stroke'-parameter present = outline
      gg: {shape: "circle", fill: "#ff0000"},          // Galaxy cluster
      g:  {shape: "ellipse", fill: "#ff0000"},         // Generic galaxy
      s:  {shape: "ellipse", fill: "#ff0000"},         // Spiral galaxy
      s0: {shape: "ellipse", fill: "#ff0000"},         // Lenticular galaxy
      sd: {shape: "ellipse", fill: "#ff0000"},         // Dwarf galaxy
      e:  {shape: "ellipse", fill: "#ff0000"},         // Elliptical galaxy
      i:  {shape: "ellipse", fill: "#ff0000"},         // Irregular galaxy
      oc: {shape: "circle", fill: "#ffcc00", stroke: "#ffcc00", width: 1.5},             // Open cluster
      gc: {shape: "circle", fill: "#ff9900"},          // Globular cluster
      en: {shape: "square", fill: "#ff00cc"},          // Emission nebula
      bn: {shape: "square", fill: "#ff00cc", stroke: "#ff00cc", width: 2},               // Generic bright nebula
      sfr:{shape: "square", fill: "#cc00ff", stroke: "#cc00ff", width: 2},               // Star forming region
      rn: {shape: "square", fill: "#00ooff"},          // Reflection nebula
      pn: {shape: "diamond", fill: "#00cccc"},         // Planetary nebula
      snr:{shape: "diamond", fill: "#ff00cc"},         // Supernova remnant
      dn: {shape: "square", fill: "#999999", stroke: "#999999", width: 2},               // Dark nebula grey
      pos:{shape: "marker", fill: "#cccccc", stroke: "#cccccc", width: 1.5}              // Generic marker
    }
  },
  planets: {  //Show planet locations, if date-time is set
    show: true, // List of all objects to show
    which: ["sol", "mer", "ven", "ter", "lun", "mar", "jup", "sat", "ura", "nep"],
    // Font styles for planetary symbols
    style: { fill: "#00ccff", font: "bold 17px 'Lucida Sans Unicode', Consolas, sans-serif",
             align: "center", baseline: "middle" },
    symbols: {  // Character and color for each symbol in 'which', simple circle \u25cf
      "sol": {symbol: "\u2609", fill: "#ffff00"},
      "mer": {symbol: "\u263f", fill: "#cccccc"},
      "ven": {symbol: "\u2640", fill: "#eeeecc"},
      "ter": {symbol: "\u2295", fill: "#00ffff"},
      "lun": {symbol: "\u25cf", fill: "#ffffff"}, // overridden by generated cresent
      "mar": {symbol: "\u2642", fill: "#ff9999"},
      "cer": {symbol: "\u26b3", fill: "#cccccc"},
      "ves": {symbol: "\u26b6", fill: "#cccccc"},
      "jup": {symbol: "\u2643", fill: "#ff9966"},
      "sat": {symbol: "\u2644", fill: "#ffcc66"},
      "ura": {symbol: "\u2645", fill: "#66ccff"},
      "nep": {symbol: "\u2646", fill: "#6666ff"},
      "plu": {symbol: "\u2647", fill: "#aaaaaa"},
      "eri": {symbol: "\u25cf", fill: "#eeeeee"}
    }
  },
  constellations: {
    show: true,    // Show constellations
    names: true,   // Show constellation names
    desig: true,   // Show short constellation names (3 letter designations)
    namestyle: { fill:"#cccc99", align: "center", baseline: "middle",
                 font: ["14px Helvetica, Arial, sans-serif",  // Style for constellations
                        "12px Helvetica, Arial, sans-serif",  // Different fonts for diff.
                        "11px Helvetica, Arial, sans-serif"]},// ranked constellations
    lines: true,   // Show constellation lines, style below
    linestyle: { stroke: "#cccccc", width: 1, opacity: 0.6 },
    bounds: true, // Show constellation boundaries, style below
    boundstyle: { stroke: "#cccc00", width: 0.5, opacity: 0.8, dash: [2, 4] }
  }, 
  mw: {
    show: false,     // Show Milky Way as filled multi-polygon outlines
    style: { fill: "#ffffff", opacity: 0.15 }  // Style for MW layers
  },
  lines: {  // Display & styles for graticule & some planes
    graticule: { show: false, stroke: "#cccccc", width: 0.6, opacity: 0.8,  
      // grid values: "outline", "center", or [lat,...] specific position
      lon: {pos: [""], fill: "#eee", font: "10px Helvetica, Arial, sans-serif"},
      // grid values: "outline", "center", or [lon,...] specific position
      lat: {pos: [""], fill: "#eee", font: "10px Helvetica, Arial, sans-serif"}},   
    equatorial: { show: false, stroke: "#aaaaaa", width: 1.3, opacity: 0.7 }, 
    ecliptic: { show: true, stroke: "#00d803", width: 3, opacity: 0.9 },    
    galactic: { show: false, stroke: "#cc6666", width: 1.3, opacity: 0.7 },   
    supergalactic: { show: false, stroke: "#cc66cc", width: 1.3, opacity: 0.7 }
  },
  background: {        // Background style
    fill: "#000000",   // Area fill
    opacity: 0.5,
    stroke: "#000000", // Outline
    width: 1.5
  },
  horizon: {  //Show horizon marker, if location is set and map projection is all-sky
    show: false,
    stroke: "#000099", // Line
    width: 1.0,
    fill: "#000000",   // Area below horizon
    opacity: 0.5
  }
};
 
var jsonLine = {"type":"FeatureCollection",
  // this is an array, add as many objects as you want
  "features":[
    {"type":"Feature",
     "id":"SummerTriangle",
     "properties": {
       // Name
       "n":"cPANSTARRS)",
       // Location of name text on the map
       "loc": [-67.5, 52]
     }, "geometry":{
       // the line object as an array of point coordinates,
       // always as [ra -180..180 degrees, dec -90..90 degrees]
       "type":"MultiLineString",
       "coordinates":[[[352.57, -18.7183333333], [353.83750000000003, -17.5883333333], [355.18, -16.4416666667], [356.59499999999997, -15.28], [358.0725, -14.1016666667], [359.61249999999995, -12.91], [1.2075, -11.7083333333], [2.8525, -10.495], [4.5475, -9.27333333333], [6.2875, -8.04666666667], [8.07, -6.81333333333], [9.8925, -5.575], [11.755, -4.335], [13.6525, -3.095], [15.584999999999999, -1.855], [17.549999999999997, -0.618333333333], [19.547500000000003, 0.615], [21.577499999999997, 1.84166666667], [23.635, 3.06], [25.7225, 4.27], [27.839999999999996, 5.46833333333], [29.982499999999998, 6.65333333333], [32.155, 7.82333333333], [34.355, 8.97833333333], [36.5775, 10.115], [38.8275, 11.23], [41.1025, 12.325], [43.400000000000006, 13.3966666667], [45.7225, 14.445], [48.065000000000005, 15.4666666667], [50.43, 16.46], [52.8125, 17.425], [55.215, 18.36], [57.635, 19.2633333333], [60.07000000000001, 20.1366666667], [62.51499999999999, 20.9766666667], [64.9725, 21.785], [67.435, 22.5583333333], [69.905, 23.2983333333], [72.3775, 24.0066666667], [74.8475, 24.6816666667], [77.3125, 25.3233333333], [79.77, 25.9333333333], [82.21249999999999, 26.515], [84.6375, 27.0666666667], [87.045, 27.5916666667], [89.4225, 28.0933333333], [91.7675, 28.5716666667], [94.0775, 29.0316666667], [96.3425, 29.475], [98.5625, 29.905], [100.7275, 30.325], [102.83, 30.7416666667], [104.865, 31.1566666667], [106.825, 31.575], [108.7025, 32.0016666667], [110.49249999999999, 32.4416666667], [112.1825, 32.8966666667], [113.7675, 33.375], [115.235, 33.88], [116.575, 34.4133333333], [117.78250000000001, 34.98], [118.845, 35.5816666667], [119.75, 36.2216666667], [120.49000000000001, 36.8966666667], [121.05250000000001, 37.605], [121.43249999999999, 38.3433333333], [121.62250000000002, 39.1066666667], [121.6175, 39.8833333333], [121.41499999999999, 40.6666666667], [121.0225, 41.44], [120.44500000000001, 42.1883333333], [119.7, 42.8983333333], [118.8125, 43.5533333333], [117.80499999999999, 44.1366666667]  ]]
}
}
]
};
 
var jsonDate = ["19961227", "199711", "199716", "1997111", "1997116", "1997121", "1997126", "1997131", "199725", "1997210", "1997215", "1997220", "1997225", "199732", "199737", "1997312", "1997317", "1997322", "1997327", "199741", "199746", "1997411", "1997416", "1997421", "1997426", "199751", "199756", "1997511", "1997516", "1997521", "1997526", "1997531", "199765", "1997610", "1997615", "1997620", "1997625", "1997630", "199775", "1997710", "1997715", "1997720", "1997725", "1997730", "199784", "199789", "1997814", "1997819", "1997824", "1997829", "199793", "199798", "1997913", "1997918", "1997923", "1997928", "1997103", "1997108", "19971013", "19971018", "19971023", "19971028", "1997112", "1997117", "19971112", "19971117", "19971122", "19971127", "1997122", "1997127", "19971212", "19971217", "19971222", "19971227", "199811"] ;
 
//var dd;
//$.get("ephfDate.json", function(data) {
//  dd = data;
//}, 'json');
 
// Line style for comet track
var lineStyle = {
  stroke:"#f44242",
  fill: "rgba(255, 204, 204, 0.4)",
  width:5
};
 
// Text style for comet track name
var textStyle = {
  fill:"#ff0000",
  font: "bold 40px Helvetica, Arial, sans-serif",
  align: "center",
  baseline: "bottom"
};
 
// Text style points along track
var textStyleTrack = {
  fill:"#9111db",
  font: "bold 10px Helvetica, Arial, sans-serif",
  align: "left",
  baseline: "top"
};
 
 
// Use the add function to read and plot comet track
// from the generated ephf.json file
//Celestial.add({ file:'ephf.json', type:'line', callback: function(error, json) {
Celestial.add({ type:'line', callback: function(error, json) {
  if (error) return console.warn(error);
 
  // Load the geoJSON file and transform to correct coordinate system, if necessary
  var asterism = Celestial.getData(jsonLine, config.transform);
  //var asterism = Celestial.getData(json, config.transform);
 
  // Add to celestial objects container in d3
  Celestial.container.selectAll(".asterisms")
                                .data(asterism.features)
                                .enter().append("path")
                                .attr("class", "ast");
 
  // Trigger redraw to display changes
  Celestial.redraw();
 
}, redraw: function() {
 
  // Select the added objects by class name as given previously
  Celestial.container.selectAll(".ast").each(function(d) {
 
    Celestial.setTextStyle(textStyleTrack);
 
    pts = d.geometry.coordinates;
 
    //alert(pts.length);
    for (var i=0; i < 75; i++) {
 
      //alert(pts[0][i]);
      pt = Celestial.mapProjection(pts[0][i]);
      //alert(pts[0][10]);
      //alert(pts[0]);
      //alert(pts[1]);
      Celestial.context.fillText(jsonDate[i],pt[0],pt[1]);
 
    }
 
    // Set line styles
    Celestial.setStyle(lineStyle);
 
    // Project objects on map
    Celestial.map(d);
 
    // Draw on canvas
    //Celestial.context.fill();
    Celestial.context.stroke();
   
    // If point is visible (this doesn't work automatically for points)
    if (Celestial.clip(d.properties.loc)) {
     
      // get point coordinates
      pt = Celestial.mapProjection(d.properties.loc);
      //alert(pt);
      // Set text styles      
      Celestial.setTextStyle(textStyle);
 
      // and draw text on canvas
      Celestial.context.fillText(d.properties.n, pt[0], pt[1]);
 
    }
  });        
}
 
});
 
// Display map with the configuration above or any subset thereof
Celestial.display(config);
 
</script>
 
 
<div id="celestial-map">
</div>
 
<div id="celestial-form">
</div>
 
</body>
 
</html>

**Comment by puziip** [comment]
I thought it was something very silly. After trying to embed the map in another webpage and finding the canvas element, for the map, in the DOM was outside the div labelled 'celestial-map' I wandered why as none of your examples had this. I have just realised that in the config I had 'container: '#celestial-map' ". As a web developer my default when references id's is to add the #. Taking this out gave me the form. Silly mistake - sorry to waste your time!

Paresh
